docker/archive/archive_dest.go

```diff
@@ -0,0 +1,224 @@
+package archive
+
+import (
+	"archive/tar"
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/containers/image/manifest"
+	"github.com/containers/image/types"
+)
+
+type archiveImageDestination struct {
+	ref       archiveReference
+	tar       *tar.Writer
+	writer    io.WriteCloser
+	committed bool // writer has been closed
+}
+
+// TODO: Refactor this so we can implement it as an io.Writer and share the
+// interface between this and docker-archive.
+func newImageDestination(ctx *types.SystemContext, ref archiveReference) (types.ImageDestination, error) {
+	fh, err := os.OpenFile(string(ref), os.O_WRONLY|os.O_EXCL|os.O_CREATE, 0644)
+	if err != nil {
+		// FIXME: It should be possible to modify archives, but the only really
+		//        sane way of doing it is to create a copy of the image, modify it
+		//        and then do a rename(2).
+		if os.IsExist(err) {
+			err = fmt.Errorf("docker-archive doesn't support modifying existing images")
+		}
+		return nil, err
+	}
+
+	return &archiveImageDestination{
+		ref:       ref,
+		tar:       tar.NewWriter(fh),
+		writer:    fh,
+		committed: false,
+	}, nil
+}
+
+// Reference returns the reference used to set up this destination.  Note that this should directly correspond to user's intent,
+// e.g. it should use the public hostname instead of the result of resolving CNAMEs or following redirects.
+func (d *archiveImageDestination) Reference() types.ImageReference {
+	return d.ref
+}
+
+// Close removes resources associated with an initialized ImageDestination, if any.
+func (d *archiveImageDestination) Close() {
+	d.writer.Close()
+	// FIXME: We don't clean up here if we errored out...
```

Comment:
? It should be the responsibility of the `ImageDestination` user to always call `Close`.